### PR TITLE
Feature/sensordownloads

### PR DESCRIPTION
**Pull request by CalebSchwartz**
Sensor Download API calls
Writing API calls for sensor downloads
- [x] Enhancement
- [x] Updated unit tests

#### Unit test coverage
****working on getting these to pass with coverage, currently all are passing without but not with. Unsure why.
```shell
COVERAGE RESULTS SHOULD BE POSTED HERE
PLEASE REVIEW CONTRIBUTING.md FOR MORE
DETAILS ON GENERATING COVERAGE RESULTS
```
image
image

#### Bandit analysis
```bandit -r src
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    running on Python 3.9.2
Run started:2021-04-02 15:28:09.956640

Test results:
        No issues identified.

Code scanned:
        Total lines of code: 6494
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
        Total issues (by confidence):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
Files skipped (0):
```
## Added features and functionality
added functionality to interact with sensor_download endpoint

### CODE REVIEW

**Comment by jshcodes** [comment]
Workflow failures aren't your fault.  Think it has to do with the workflow having a hard time getting to the repo secret.  I will kick off another run after this one fails out and we can re-confirm.

**Comment by CalebSchwartz** [comment]
thanks! do you have a test environment contributors can utilize? I'm currently using my companies test tenant we have to make API calls but its not exactly "to policy" :)

**Comment by jshcodes** [comment]
Not yet, but this is an excellent idea.  Let me see if I can't work on something.

So the unit test failures have to do with this being in your fork, so again, a repo secrets issue. (You can create secrets that match your CID and put them in your local fork's secrets, but that's not terribly elegant.)

I have a couple of minor tweaks for this and then I think what we can do is merge it in and re-test it.

File: src/falconpy/sensor_download.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                                   )
+        return returned
+
+    def DownloadSensorInstallerById(self: object, _id: str, file_name: str, download_path: str = "sensor_downloads"):
+        """
+        download the sensor by the sha256 into the specified directory.
+        the path will be created for the user if it does not already exist
+        """
+        # create the directory if doesn't exist
+        os.makedirs(download_path, exist_ok=True)
```
Comment (jshcodes) [inline src/falconpy/sensor_download.py:31]:
```suggestion

```

File: src/falconpy/sensor_download.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        returned = service_request(caller=self,
+                                   method="GET",
+                                   endpoint=FULL_URL,
+                                   headers=HEADERS,
+                                   verify=self.ssl_verify
+                                   )
+        # write the newly downloaded sensor into the aforementioned directory with provided file name
+        with open(os.path.join(download_path, file_name), "wb") as sensor:
+            sensor.write(returned)
+        return True if returned else False
```
Comment (jshcodes) [inline src/falconpy/sensor_download.py:44]:
```suggestion
        if download_path:
             # create the directory if doesn't exist
             os.makedirs(download_path, exist_ok=True)
             # write the newly downloaded sensor into the aforementioned directory with provided file name
             with open(os.path.join(download_path, file_name), "wb") as sensor:
                    sensor.write(returned)
                    # Let's add a "successfully downloaded" response here, below does not match pattern just an example
                    returned = { "message": "Download successful" }
        
        return returned
```

File: src/falconpy/sensor_download.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        returned = service_request(caller=self,
+                                   method="GET",
+                                   endpoint=FULL_URL,
+                                   params=PARAMS,
+                                   headers=HEADERS,
+                                   verify=self.ssl_verify
+                                   )
+        return returned
+
+    def DownloadSensorInstallerById(self: object, _id: str, file_name: str, download_path: str = "sensor_downloads"):
```
Comment (jshcodes) [inline src/falconpy/sensor_download.py:25]:
```suggestion
    def DownloadSensorInstallerById(self: object, _id: str, file_name: str, download_path: str = None) -> object:
```

File: src/falconpy/sensor_download.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                                   endpoint=FULL_URL,
+                                   headers=HEADERS,
+                                   verify=self.ssl_verify
+                                   )
+        # write the newly downloaded sensor into the aforementioned directory with provided file name
+        with open(os.path.join(download_path, file_name), "wb") as sensor:
+            sensor.write(returned)
+        return True if returned else False
+
+    def GetSensorInstallersEntities(self: object, ids: list):
```
Comment (jshcodes) [inline src/falconpy/sensor_download.py:46]:
```suggestion
    def GetSensorInstallersEntities(self: object, ids: list or str) -> dict:
```

File: src/falconpy/sensor_download.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        HEADERS = self.headers
+        returned = service_request(caller=self,
+                                   method="GET",
+                                   endpoint=FULL_URL,
+                                   headers=HEADERS,
+                                   verify=self.ssl_verify
+                                   )
+        return returned
+
+    def GetSensorInstallersCCIDByQuery(self: object):
```
Comment (jshcodes) [inline src/falconpy/sensor_download.py:62]:
```suggestion
    def GetSensorInstallersCCIDByQuery(self: object) -> dict:
```

File: src/falconpy/sensor_download.py (hunk 3 above)
Comment (jshcodes) [reply src/falconpy/sensor_download.py:25]: Curious why you decided to go with a private for _id as opposed to using the ids pattern?  Is it because it's always one ID?

File: src/falconpy/sensor_download.py (hunk 1 above)
Comment (jshcodes) [reply src/falconpy/sensor_download.py:31]: Moving this down below.

File: src/falconpy/sensor_download.py (hunk 2 above)
Comment (jshcodes) [reply src/falconpy/sensor_download.py:44]: I don't think we should have the library downloading files by default.  The default functionality should match other classes which is to return the object as it's received from the API.

That being said, this is really cool, and I bet a lot of folks would use it this way, so let's keep it as an optional variation.

We need to craft a response for the developer.  We might get away with using generate_error_result and sending a 200 for the code.

File: src/falconpy/sensor_download.py (hunk 4 above)
Comment (jshcodes) [reply src/falconpy/sensor_download.py:46]: Adding decorations

File: src/falconpy/sensor_download.py (hunk 5 above)
Comment (jshcodes) [reply src/falconpy/sensor_download.py:62]: Adding decorations

File: tests/test_sensor_download.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        assert self._get_metadata_for_filter() == True
+
+    def test_get_ccid(self):
+        assert self._get_cid() == True
+
+    def test_get_shas(self):
+        assert len(self._get_multiple_shas()) > 0
+
+    def test_get_mutliple_shas(self):
+        assert self._get_metadata_for_ids() == True
```
Comment (jshcodes) [inline tests/test_sensor_download.py:60]:
```suggestion
        assert self._get_metadata_for_ids() == True
        
```

Comment (jshcodes) [reply tests/test_sensor_download.py:60]: Linting fix

File: src/falconpy/sensor_download.py (hunk 2 above)
Comment (jshcodes) [reply src/falconpy/sensor_download.py:44]: Spacing might be borked on this suggestion, the GH editor wouldn't let me use tab...

File: src/falconpy/sensor_download.py (hunk 3 above)
Comment (CalebSchwartz) [reply src/falconpy/sensor_download.py]: flipped in newest commit, been writing personal code calling things like that as no one else uses it :) changed to be id_

File: src/falconpy/sensor_download.py (hunk 2 above)
Comment (CalebSchwartz) [reply src/falconpy/sensor_download.py]: updated this a bit. may want to request API team to return a status code along with the downloads. for now checking that user passes a filename, download directory and the request returns the bytes.

File: src/falconpy/sensor_download.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
+from ._util import service_request, parse_id_list, generate_ok_result
+from ._service_class import ServiceClass
+
+import os
+
+
+class Sensor_Download(ServiceClass):
+
+    def GetCombinedSensorInstallersByQuery(self: object, params: dict) -> dict:
```
Comment (jshcodes) [inline src/falconpy/sensor_download.py:9]:
```suggestion
    def GetCombinedSensorInstallersByQuery(self: object, params: dict = {}) -> dict:
```

Comment (jshcodes) [reply src/falconpy/sensor_download.py:9]: Parameter payload is not required for this operation

File: src/falconpy/sensor_download.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+                                   )
+        if file_name and download_path and isinstance(returned, bytes):
+            os.makedirs(download_path, exist_ok=True)
+            # write the newly downloaded sensor into the aforementioned directory with provided file name
+            with open(os.path.join(download_path, file_name), "wb") as sensor:
+                sensor.write(returned)
+            returned = generate_ok_result(message="Download successful")
+        return returned
+
+    def GetSensorInstallersEntities(self: object, ids: list) -> object:
```
Comment (jshcodes) [inline src/falconpy/sensor_download.py:47]:
```suggestion
    def GetSensorInstallersEntities(self: object, ids: list or str) -> object:
```

Comment (jshcodes) [reply src/falconpy/sensor_download.py:47]: parse_id_list allows us to accept strings or lists for the ids array.

File: src/falconpy/sensor_download.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        HEADERS = self.headers
+        returned = service_request(caller=self,
+                                   method="GET",
+                                   endpoint=FULL_URL,
+                                   headers=HEADERS,
+                                   verify=self.ssl_verify
+                                   )
+        return returned
+
+    def GetSensorInstallersByQuery(self: object, params: dict) -> dict:
```
Comment (jshcodes) [inline src/falconpy/sensor_download.py:77]:
```suggestion
    def GetSensorInstallersByQuery(self: object, params: dict = {}) -> dict:
```

Comment (jshcodes) [reply src/falconpy/sensor_download.py:77]: Parameters payload is not required for this operation.

File: src/falconpy/_util.py [hunk 10]
```diff
@@ -190,3 +190,7 @@ def perform_request(method: str = "", endpoint: str = "", headers: dict = None,
 def generate_error_result(message: str = "An error has occurred. Check your payloads and try again.", code: int = 500) -> dict:
     """ Normalized error messaging handler. """
     return Result()(status_code=code, headers={}, body={"errors": [{"message": f"{message}"}], "resources": []})
+
+  
+def generate_ok_result(message: str = "Request returned with success", code: int = 200) -> dict:
```
Comment (jshcodes) [inline src/falconpy/_util.py:195]: Excellent. This is the right way to handle this.

File: tests/test_sensor_download.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    @staticmethod
+    def _get_cid():
+        resp = sensor_download_client.GetSensorInstallersCCIDByQuery()
+        return True if resp["status_code"] in AllowedResponses else False
+
+    @staticmethod
+    def _get_multiple_shas():
+        params = {"filter": 'platform:"windows"', "sort": "release_date|desc"}
+        shas = sensor_download_client.GetSensorInstallersByQuery(params)["body"].get("resources", ["b7dbe32495c6e23ca2e04784b98d5a1770163b9f42524cafc062292b533f03c9"])
+        return shas
```
Comment (jshcodes) [inline tests/test_sensor_download.py:26]:
```suggestion
    def _get_multiple_shas():
        params = {"filter": 'platform:"windows"', "sort": "release_date|desc"}
        shas = sensor_download_client.GetSensorInstallersByQuery(params)["body"].get("resources", [""])
        return shas
```

Comment (jshcodes) [reply tests/test_sensor_download.py:26]: We need to come up with a way to retrieve this value dynamically, or we'll be stuck updating the unit test as the sha changes.

File: tests/test_sensor_download.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+    def _download_sensor(self):
+        file_name = "falconwinddows.exe"
+        directory_path = "sensor_downloads"
+        sha_id = self._get_multiple_shas()[0]
+        resp = sensor_download_client.DownloadSensorInstallerById(_id=sha_id, file_name=file_name, download_path=directory_path)
+        return resp
```
Comment (jshcodes) [inline tests/test_sensor_download.py:33]: This unit test is a little strange but I think I understand how it works. From what I'm reading here we're returning the value of **resp** which in this case would contain the binary for the download.  Python would calculate that to equal a _True_ that the assert statement is expecting down below.  (Or am I misreading?)

**Review by jshcodes — CHANGES_REQUESTED** [review]

File: tests/test_sensor_download.py (hunk 11 above)
Comment (CalebSchwartz) [reply tests/test_sensor_download.py:26]: if the API returns a 429, we could just retry but dont like doing that in tests typically. that what was causing issues when i was testing, was getting constant 429s so just hard coded one.

I was thinking of reworking tests to utilize  pytest-httpserver but project for another weekend I think. Would need way to get swagger docs into it

**Review by jshcodes — APPROVED** [review]

**Review by jshcodes — APPROVED** [review]
